### PR TITLE
mass qdel proc to replace qdel_in(), batching all qdel timer into a single timer

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
mass qdel proc to replace qdel_in(), batching all qdel timer into a single timer

Bacon didn't like #10178
but I don't like this in fact

This PR is to show the code difference.

## Why It's Good For The Game
performacne

## Testing Photographs and Procedure

Mainly the same to
* #10178

image

## Changelog
:cl:
code: qdel_in() no longer uses addtimer() for each item to qdel. It sends an item to a list with scheduled time assoc item list.
/:cl:

### CODE REVIEW

File: code/__HELPERS/qdel.dm [hunk 1]
```diff
@@ -8,3 +9,62 @@
...
+	return timer_key
+
+// Internal usage only. Timer subsystem will call it.
+/proc/_mass_qdel(timer_key)
+	var/list/target_entry = GLOB.qdel_timers[timer_key]
+	while(length(target_entry)) // fun fact: length(LIST) is faster than LIST.len
+		var/datum/qdel_target = target_entry[length(target_entry)]
+		if(!QDELETED(qdel_target))
+			qdel(qdel_target)
+		target_entry.len-- // fun fact: this removes the last item from the list. convenient.
```
Comment (PowerfulBacon) [inline code/__HELPERS/qdel.dm]: What's the point of altering the size of the list? Just do a for loop

File: code/__HELPERS/qdel.dm [hunk 2]
```diff
@@ -8,3 +9,62 @@
...
+	if(!length(GLOB.qdel_timers[timer_key]))
+		GLOB.qdel_timers -= timer_key
+	return
+
+/proc/qdel_timer_cancle(datum/item, timer_key)
```
Comment (PowerfulBacon) [inline code/__HELPERS/qdel.dm]: This is not how you spell cancel

**Comment by PowerfulBacon** [comment]
Not a huge fan of this PR, seems like unnecessary jank to automate bad code.

**Comment by EvilDragonfiend** [comment]
> Not a huge fan of this PR, seems like unnecessary jank to accomodate bad code, if something wants to delete 100000 items in 1 tick, then it should handle that internally.

It wouldn't be 100,000 items in a tick, but `qdel_in()` itself is already expensive because it uses `addtimer()`, but also there's no way to handle this internally because `qdel_in()` is massively used in everywhere.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
This idea is okay, but it has no way to deal with object deletions and the implementation of timers under the hood is actually pretty efficient and effectively is just looping through a list anyway. I don't think this is necessary and I'm not convinced that this will provide large performance improvements.

The idea behind the PR is something that I spent a day coding before realising that the timer implementation was already really good for dealing with a lot of timers and that actually testing for improvement is very difficult.

**Comment by EvilDragonfiend** [comment]
> This idea is okay, but it has no way to deal with object deletions and the implementation of timers under the hood is actually pretty efficient and effectively is just looping through a list anyway. I don't think this is necessary and I'm not convinced that this will provide large performance improvements.

The issue doesn't come from how SStimer is coded. The issue comes from how the timer system is used to qdel massive amount of stuff. For example, lighting flickering is one of examples that spams qdel timer.

**Comment by PowerfulBacon** [comment]
The biggest difference isn't performance, its that the subsystem can't be TICK CHECKED mid-execution which will result in things being deleted on the same tick (rather than being deferred by timers) but it will also allow the subsystem to overrun its tick allocation to do that

**Comment by Rukofamicom** [comment]
> the subsystem can't be TICK CHECKED mid-execution

Nope 🔑

**Pull request closed without merging**